File: src/mochi_websocket_ffi.erl
-module(mochi_websocket_ffi).
-export([parse_json/1, get_field/2, get_string/1, get_dict/1]).

%% Parse JSON string to Erlang term
parse_json(JsonBin) when is_binary(JsonBin) ->
    try
        case json:decode(JsonBin) of
            Map when is_map(Map) -> {ok, Map};
            Other -> {ok, Other}
        end
    catch
        _:_ -> {error, <<"Invalid JSON">>}
    end.

%% Get a field from a map
get_field(Map, Field) when is_map(Map), is_binary(Field) ->
    case maps:find(Field, Map) of
        {ok, Value} -> {ok, Value};
        error -> {error, nil}
    end;
get_field(_, _) ->
    {error, nil}.

%% Get string value
get_string(Bin) when is_binary(Bin) ->
    {ok, Bin};
get_string(_) ->
    {error, nil}.

%% Get dict/map value
get_dict(Map) when is_map(Map) ->
    {ok, Map};
get_dict(_) ->
    {error, nil}.

File: src/mochi_websocket_ffi.mjs
// WebSocket FFI for JavaScript target
import { Ok, Error } from "./gleam.mjs";
import { toList } from "./gleam.mjs";

export function parse_json(jsonStr) {
  try {
    const parsed = JSON.parse(jsonStr);
    return new Ok(parsed);
  } catch (e) {
    return new Error("Invalid JSON");
  }
}

export function get_field(obj, field) {
  if (obj && typeof obj === "object" && field in obj) {
    return new Ok(obj[field]);
  }
  return new Error(undefined);
}

export function get_string(value) {
  if (typeof value === "string") {
    return new Ok(value);
  }
  return new Error(undefined);
}

export function get_dict(value) {
  if (value && typeof value === "object" && !Array.isArray(value)) {
    return new Ok(value);
  }
  return new Error(undefined);
}
